### PR TITLE
backend: add prometheus metric for large snapshot duration.

### LINKED ISSUE #7878: metric: create Prometheus metric for snapshotting duration.

### PR DESCRIPTION
**Pull request by fanminshi**
FIXES #7878

### CODE REVIEW

**Comment by fanminshi** [comment]
I manually create 2 snapshots which take 10 seconds each.

I saw following metrics:

```
etcd_disk_backend_snapshot_duration_seconds_bucket{le="1"} 0
etcd_disk_backend_snapshot_duration_seconds_bucket{le="2"} 0
etcd_disk_backend_snapshot_duration_seconds_bucket{le="4"} 0
etcd_disk_backend_snapshot_duration_seconds_bucket{le="8"} 0
etcd_disk_backend_snapshot_duration_seconds_bucket{le="16"} 2
etcd_disk_backend_snapshot_duration_seconds_bucket{le="32"} 2
etcd_disk_backend_snapshot_duration_seconds_bucket{le="64"} 2
etcd_disk_backend_snapshot_duration_seconds_bucket{le="128"} 2
etcd_disk_backend_snapshot_duration_seconds_bucket{le="256"} 2
etcd_disk_backend_snapshot_duration_seconds_bucket{le="512"} 2
etcd_disk_backend_snapshot_duration_seconds_bucket{le="+Inf"} 2
etcd_disk_backend_snapshot_duration_seconds_sum 20.001328967
etcd_disk_backend_snapshot_duration_seconds_count 2
```

I am unsure why buckets 32-512 also have counts 2.

File: mvcc/backend/metrics.go [hunk 1]
```diff
@@ -24,8 +24,18 @@ var (
 		Help:      "The latency distributions of commit called by backend.",
 		Buckets:   prometheus.ExponentialBuckets(0.001, 2, 14),
 	})
+
+	snapShotDurations = prometheus.NewHistogram(prometheus.HistogramOpts{
```
Comment (heyitsanthony) [inline mvcc/backend/metrics.go]: `snapshotDurations`

File: mvcc/backend/metrics.go [hunk 2]
```diff
@@ -24,8 +24,18 @@ var (
...
+		Namespace: "etcd",
+		Subsystem: "disk",
+		Name:      "backend_snapshot_duration_seconds",
+		Help:      "The latency distributions of Snapshot called by backend.",
+		// 1 second -> 1024 seconds
+		Buckets: prometheus.ExponentialBuckets(1, 2, 10),
```
Comment (heyitsanthony) [inline mvcc/backend/metrics.go]: 1024 seconds is extreme, probably want to capture something like [10ms -- 30s]

File: mvcc/backend/metrics.go [hunk 3]
```diff
@@ -24,8 +24,18 @@ var (
 		Help:      "The latency distributions of commit called by backend.",
 		Buckets:   prometheus.ExponentialBuckets(0.001, 2, 14),
 	})
+
+	snapShotDurations = prometheus.NewHistogram(prometheus.HistogramOpts{
+		Namespace: "etcd",
+		Subsystem: "disk",
+		Name:      "backend_snapshot_duration_seconds",
+		Help:      "The latency distributions of Snapshot called by backend.",
```
Comment (heyitsanthony) [inline mvcc/backend/metrics.go]: `The latency distribution of backend snapshots.`

File: mvcc/backend/metrics.go [hunk 4]
```diff
@@ -24,8 +24,18 @@ var (
...
+		// 1 second -> 512 seconds
+		Buckets: prometheus.ExponentialBuckets(1, 2, 10),
```
Comment (heyitsanthony) [inline mvcc/backend/metrics.go]: probably want something like 10ms -- 1minute. 512 seconds is a lot

Comment (fanminshi) [inline mvcc/backend/metrics.go]: @xiang90 wants to track large snapshot duration fro 1 second to around 10 min.

Comment (heyitsanthony) [inline mvcc/backend/metrics.go]: ok...

Comment (xiang90) [inline mvcc/backend/metrics.go]: we can track low numbers that are common cases. when cluster starts to suffer we probably want to track large numbers. if the snap is 8gb with slow network 100s seconds is possible

Comment (fanminshi) [inline mvcc/backend/metrics.go]: okay, should we do 10s to 10mins?

**Comment by fanminshi** [comment]
@xiang90 @heyitsanthony.

How about 10ms to 10 mins?

**Comment by wenjiaswe** [comment]
@gyuho I didn't find this one in [CHANGELOG-3.2.md](https://github.com/coreos/etcd/blob/master/CHANGELOG-3.2.md), it is merged in 3.2 right? Just want to confirm, I can add in the changelog and backport to 3.1 if it's safe.

**Comment by gyuho** [comment]
@wenjiaswe Yeah, it's in 3.2 not in 3.1 Let's backport and update changelog.

**Pull request merged**